tests/test-dladdr.c

```diff
@@ -0,0 +1,183 @@
+/* On Unix like os compile with "-Wl,--export-dynamic -fpie" (default with cmake) */
+
+/* required for non Windows builds, must be set in front of the first system include */
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int verbose = 0;
+
+typedef enum {
+    Pass = 1,
+    PassWithoutSymbol = 2,
+    PassWithDifferentAddress = 3,
+    Fail = 0,
+    NoInfo = -1,
+} ExpectedResult;
+
+typedef void (* func) (void);
+
+static void print_dl_info( Dl_info *info, char *prefix, char *suffix )
+{
+    printf( "%sfilename: %s base: %p symbol name: '%s' addr: %p%s", prefix, info->dli_fname, info->dli_fbase, info->dli_sname, info->dli_saddr, suffix );
+}
+
+/**
+ * @brief check information returned by dladdr
+ * @param hint text describing what to test
+ * @param addr address to check
+ * @param addrsym
+ * @param expected check against expected result
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr( const char *hint, void *addr, char *addrsym, ExpectedResult expected_result )
+{
+    Dl_info info;
+    int result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match && !info.dli_sname)
+                 || (expected_result == PassWithDifferentAddress && sym_match && !addr_match)
+                 || (expected_result == Fail && (!sym_match || !addr_match));
+    }
+    printf( "checking '%s' - address %p which should have symbol '%s' -> %s%s", hint, addr, addrsym, passed ? "passed" : "failed", verbose || !passed ? " " : "\n" );
+    if( verbose || !passed )
+    {
+        if( !result )
+            printf( "(could not get symbol information for address %p)\n", addr );
+        else
+            print_dl_info( &info, "(", ")\n");
+    }
+    return !passed;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
```

Comment:
After calling dlclose() `addr` is not valid anymore.

Reply:
But if the executable in question is linked with that library, the address should point to the correct location into the process memory, regardless if that library was opened second time.

Reply:
Probably yes, but in general this usage of dlsym/dlclose is not correct according to POSIX documentation.